snuba/datasets/metrics_aggregate_processor.py

```diff
@@ -0,0 +1,182 @@
+
+from snuba import environment, settings
+from snuba.consumers.types import KafkaMessageMetadata
+from snuba.processor import (
+    AggregateInsertBatch,
+    MessageProcessor,
+    ProcessedMessage,
+    _ensure_valid_date,
+)
+from snuba.query.expressions import (
+    Expression,
+    FunctionCall,
+    Literal,
+    OptionalScalarType,
+)
+from snuba.utils.metrics.wrapper import MetricsWrapper
+
+snuba_metrics = MetricsWrapper(environment.metrics, "metrics.processor")
+
+
+def _literal(value: OptionalScalarType) -> Literal:
+    return Literal(None, value)
+
+
+def _array_literal(values: Sequence[OptionalScalarType]) -> FunctionCall:
+    return FunctionCall(None, "array", tuple(map(_literal, values)))
+
+
+def _call(function_name: str, arguments: Tuple[Expression, ...]) -> FunctionCall:
+    return FunctionCall(None, function_name, arguments)
+
+
+METRICS_SET_TYPE = "s"
+METRICS_DISTRIBUTIONS_TYPE = "d"
+METRICS_COUNTERS_TYPE = "c"
+
+
+class MetricsAggregateProcessor(MessageProcessor, ABC):
+    TEN_SECONDS = 10
+    ONE_MINUTE = 60
+    ONE_HOUR = 3600
+    ONE_DAY = 3600 * 24
+    GRANULARITIES_SECONDS = [TEN_SECONDS, ONE_MINUTE, ONE_HOUR, ONE_DAY]
+
+    @abstractmethod
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+    def process_message(
+        self, message: Mapping[str, Any], metadata: KafkaMessageMetadata
+    ) -> Optional[ProcessedMessage]:
+        if not self._should_process(message):
+            return None
+
+        timestamp = _ensure_valid_date(datetime.utcfromtimestamp(message["timestamp"]))
+        assert timestamp is not None
+
+        keys = []
+        values = []
+        tags = message["tags"]
+        assert isinstance(tags, Mapping)
+        for key, value in sorted(tags.items()):
+            assert key.isdigit()
+            keys.append(int(key))
+            assert isinstance(value, int)
+            values.append(value)
+
+        processed = [
+            {
+                "org_id": _literal(message["org_id"]),
+                "project_id": _literal(message["project_id"]),
+                "metric_id": _literal(message["metric_id"]),
+                "timestamp": _call(
+                    "toStartOfInterval",
+                    (
+                        _call("toDateTime", (_literal(timestamp.isoformat()),),),
+                        _call("toIntervalSecond", (_literal(granularity),)),
+                    ),
+                ),
+                "tags.key": _array_literal(keys),
+                "tags.value": _array_literal(values),
+                **self._process_values(message),
+                "retention_days": _literal(message["retention_days"]),
+                "partition": metadata.partition,
+                "offset": metadata.offset,
```

Comment:
These are not supposed to exist in the aggregate table. I guess it still work because you filter only the columns that actually exist.